taskvine/src/manager/vine_manager.c

```diff
@@ -501,6 +505,167 @@ static int handle_transfer_hostport(struct vine_manager *q, struct vine_worker_i
+			warn(D_VINE, "it uses a version of a shared library different from the one at the worker.\n");
+			break;
+		default:
+			break;
+		}
+	}
+
+	vine_task_info_add(q, t);
+
+	debug(D_VINE,
+			"%s (%s) done in %.02lfs total tasks %lld average %.02lfs",
+			w->hostname,
+			w->addrport,
+			(t->time_when_done - t->time_when_commit_start) / 1000000.0,
+			(long long)w->total_tasks_complete,
+			w->total_task_time / w->total_tasks_complete / 1000000.0);
+
+	return;
+}
+
+static vine_result_code_t get_completion_result(struct vine_manager *q, struct vine_worker_info *w, const char *line)
+{
+	if (!q || !w || !line)
+		return VINE_WORKER_FAILURE;
+
+	struct vine_task *t;
+	int task_status, exit_status;
+	uint64_t task_id;
+	int64_t output_length;
+
+	timestamp_t execution_time, start_time, end_time;
+	timestamp_t observed_execution_time;
+
+	// Format: task completion status, exit status (exit code or signal), output length, execution time, task_id
+	int n = sscanf(line,
+			"complete %d %d %" SCNd64 " %" SCNd64 " %" SCNd64 " %" SCNd64 "",
+			&task_status,
+			&exit_status,
+			&output_length,
+			&start_time,
+			&end_time,
+			&task_id);
+
+	if (n < 6) {
+		debug(D_VINE, "Invalid message from worker %s (%s): %s", w->hostname, w->addrport, line);
+		return VINE_WORKER_FAILURE;
+	}
+
+	execution_time = end_time - start_time;
+
+	/* If the worker sent back a task we have never heard of, then discard the following data. */
+	t = itable_lookup(w->current_tasks, task_id);
+	if (!t) {
+		debug(D_VINE,
+				"Unknown task completion from worker %s (%s): no task %" PRId64
+				" assigned to worker. Ignoring result.",
+				w->hostname,
+				w->addrport,
+				task_id);
+
+		time_t stoptime = time(0) + vine_manager_transfer_time(q, w, output_length);
+		link_soak(w->link, output_length, stoptime);
+		return VINE_SUCCESS;
+	}
+
+	if (task_status != VINE_RESULT_SUCCESS) {
+		w->last_failure_time = timestamp_get();
+		t->time_when_last_failure = w->last_failure_time;
+	}
+
+	/* If the task was forsaken by the worker or couldn't exeute, it didn't really complete, so short circuit. */
+	if (task_status == VINE_RESULT_FORSAKEN) {
+		t->forsaken_count++;
+		itable_remove(q->running_table, t->task_id);
+		vine_task_set_result(t, task_status);
+		change_task_state(q, t, VINE_TASK_WAITING_RETRIEVAL);
+		return VINE_SUCCESS;
+	}
+
+	/* Update task stats for this completion. */
+	observed_execution_time = timestamp_get() - t->time_when_commit_end;
+
+	t->time_workers_execute_last =
+			observed_execution_time > execution_time ? execution_time : observed_execution_time;
+	t->time_workers_execute_last_start = start_time;
+	t->time_workers_execute_last_end = end_time;
+	t->time_workers_execute_all += t->time_workers_execute_last;
+	t->output_length = output_length;
+	t->result = task_status;
+	t->exit_code = exit_status;
+
+	/* If output is less than 1KB stdout is sent along with completion msg. retrieve it from the link. */
```

Comment:
The system will be more robust with an explicit indicator rather than an implied assumption of 1024 bytes.
Suggest you change the message so that one field indicates "how many bytes immediately follow this message" and another indicates "what is the total size of the stdout."  And then this code can act directly on that information.